{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xC001o52LoJg",
        "outputId": "f2895fbf-5de0-4430-9d57-21ebe908cf72"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting Biopython\n",
            "  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m10.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from Biopython) (1.23.5)\n",
            "Installing collected packages: Biopython\n",
            "Successfully installed Biopython-1.81\n"
          ]
        }
      ],
      "source": [
        "! pip install Biopython"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "background_save": true,
          "base_uri": "https://localhost:8080/"
        },
        "id": "mpLm8KwS6LCf",
        "outputId": "c3703ec6-4d31-43be-ac09-3e41c94660f1"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "Menu:\n",
            "1. Count k-mers from sequence\n",
            "2. Exit\n",
            "{'TTTC': 16, 'TTCA': 7, 'TCAT': 12, 'CATA': 8, 'ATAT': 21, 'TATC': 10, 'ATCA': 18, 'TCAC': 8, 'CACT': 5, 'ACTA': 7, 'CTAC': 10, 'TACC': 5, 'ACCG': 17, 'CCGT': 14, 'CGTT': 10, 'GTTT': 14, 'TTTG': 16, 'TTGA': 9, 'TGAG': 10, 'GAGT': 6, 'AGTT': 10, 'GTTC': 7, 'TTCT': 15, 'TCTT': 14, 'CTTG': 8, 'TTGT': 12, 'TGTG': 21, 'GTGC': 17, 'TGCT': 19, 'GCTG': 23, 'CTGT': 13, 'GTGT': 8, 'GTGG': 14, 'TGGA': 12, 'GGAT': 13, 'GATA': 13, 'ATAC': 7, 'TACT': 6, 'ACTC': 8, 'CTCC': 16, 'TCCT': 10, 'CCTC': 11, 'TCCC': 9, 'CCCG': 11, 'CCGA': 9, 'CGAC': 12, 'GACA': 8, 'ACAC': 40, 'CACA': 34, 'ACAA': 30, 'CAAA': 22, 'AAAG': 18, 'AAGC': 13, 'AGCC': 3, 'GCCG': 9, 'CCGC': 12, 'CGCT': 14, 'GCTC': 12, 'TCCA': 18, 'CCAT': 19, 'CATC': 20, 'TCAG': 3, 'CAGC': 12, 'GCCA': 16, 'CCAG': 13, 'AGCA': 16, 'GCAG': 13, 'CAGT': 5, 'AGTC': 5, 'GTCG': 9, 'TCGT': 10, 'CGTC': 8, 'GTCT': 12, 'TCTA': 6, 'CTAA': 8, 'TAAT': 17, 'AATC': 7, 'ATCC': 11, 'CAGA': 7, 'AGAG': 8, 'GAGA': 13, 'AGAC': 7, 'CACC': 32, 'ACCA': 17, 'CCAA': 16, 'AAAC': 12, 'AACC': 9, 'CGAA': 7, 'GAAA': 17, 'AAGA': 23, 'GACT': 9, 'ACTT': 9, 'CTTA': 8, 'TTAA': 14, 'AATT': 13, 'ATTT': 21, 'TTTA': 16, 'TTAT': 21, 'TATA': 13, 'TATT': 20, 'ATTA': 14, 'TTTT': 34, 'AATA': 10, 'ATAA': 12, 'TAAA': 11, 'AAAA': 28, 'AACA': 26, 'AAAT': 20, 'ATCT': 11, 'TTAC': 7, 'TACA': 16, 'AATG': 13, 'ATGT': 17, 'GTGA': 12, 'TGAC': 9, 'GACG': 8, 'ACGA': 12, 'GAAG': 20, 'AGAA': 15, 'AAGT': 16, 'GTTG': 6, 'TTGC': 11, 'CTGC': 18, 'CTCT': 16, 'TCTG': 19, 'CTGG': 15, 'TGGT': 9, 'GGTT': 6, 'TGTC': 11, 'TCGA': 11, 'CAAC': 22, 'AACG': 3, 'ACGG': 6, 'CGGC': 12, 'GGCT': 12, 'TGGC': 16, 'GGCA': 16, 'GCAT': 12, 'CATG': 15, 'TGCA': 15, 'GCAA': 16, 'CAAG': 21, 'AAGG': 12, 'AGGC': 9, 'GGCC': 12, 'CCGG': 11, 'CGGA': 7, 'GATT': 9, 'TGCC': 14, 'GGAG': 17, 'CGAT': 9, 'GATG': 15, 'ATGC': 13, 'CCCC': 10, 'CGCG': 5, 'GCGC': 8, 'CGCC': 10, 'CTTC': 11, 'TTCC': 13, 'CCCA': 13, 'ATCG': 12, 'ATTG': 10, 'TGGG': 9, 'GGGA': 5, 'GGAC': 6, 'ACGT': 3, 'GTCC': 6, 'GTCA': 2, 'CAGG': 6, 'AGGG': 4, 'GGGT': 4, 'GGTG': 9, 'TGAT': 14, 'ATGG': 9, 'GGTC': 5, 'TCGG': 6, 'GGGC': 7, 'AGGA': 20, 'CTCG': 11, 'CGTA': 7, 'GTAC': 10, 'TACG': 7, 'CGTG': 11, 'ATGA': 18, 'GAGG': 17, 'GGCG': 4, 'CGCA': 11, 'GAGC': 13, 'AGCG': 15, 'GCGT': 9, 'GGTA': 4, 'GTAT': 13, 'CTCA': 3, 'ACCC': 12, 'CCCT': 9, 'CCTG': 10, 'CTGA': 10, 'TGAA': 16, 'AGTA': 4, 'CATT': 13, 'GCAC': 17, 'CACG': 10, 'CGGT': 9, 'GACC': 9, 'AACT': 8, 'ACTG': 4, 'TATG': 14, 'AGAT': 16, 'GATC': 15, 'CCAC': 16, 'ACCT': 9, 'CCTT': 7, 'CAAT': 4, 'AGCT': 5, 'TGCG': 11, 'CGAG': 14, 'GCCC': 12, 'GAAC': 7, 'TGTT': 6, 'TTCG': 8, 'TCAA': 7, 'TGTA': 12, 'TCTC': 15, 'TCGC': 13, 'ACGC': 7, 'GCCT': 2, 'TCCG': 9, 'GCGA': 10, 'GTTA': 5, 'GCGG': 8, 'CTTT': 12, 'ACAT': 13, 'GCTA': 6, 'CTAT': 2, 'TTGG': 8, 'CCTA': 2, 'ACAG': 1, 'AGGT': 2, 'GGAA': 8, 'GAAT': 2, 'GCTT': 9, 'AGTG': 10, 'GTAA': 4, 'TAAG': 5, 'CGGG': 3, 'ATTC': 10, 'TAAC': 5, 'CTAG': 1, 'TAGC': 1, 'ATAG': 4, 'TAGT': 2, 'TAGA': 3, 'TTAG': 1}\n",
            "\n",
            "Menu:\n",
            "1. Count k-mers from sequence\n",
            "2. Exit\n",
            "Exiting...\n"
          ]
        }
      ],
      "source": [
        "#question 1\n",
        "from Bio import Entrez, SeqIO\n",
        "from io import StringIO\n",
        "\n",
        "# Replace with your email address (required by NCBI)\n",
        "Entrez.email = \"shyama21492@iiitd.ac.in\"\n",
        "\n",
        "s_accession_number = \"NM_001297986\"\n",
        "\n",
        "def download_sequence():\n",
        "    try:\n",
        "        # Fetch the sequence from NCBI in FASTA format\n",
        "        handle = Entrez.efetch(db=\"nucleotide\", id=s_accession_number, rettype=\"fasta\", retmode=\"text\")\n",
        "        sequence = handle.read()\n",
        "        handle.close()\n",
        "        return sequence\n",
        "    except Exception as e:\n",
        "        raise ValueError(f\"Failed to download sequence due to: {e}\")\n",
        "def count_kmers(sequence, k):\n",
        "    # Generate k-mers using zip and list comprehension\n",
        "    kmers = [''.join(chars) for chars in zip(*[sequence[i:] for i in range(k)])]\n",
        "\n",
        "    kmers_counts = {}\n",
        "    for kmer in kmers:\n",
        "        kmers_counts[kmer] = kmers_counts.get(kmer, 0) + 1\n",
        "\n",
        "    return kmers_counts\n",
        "\n",
        "def count_kmers_from_string(sequence_string, k):\n",
        "    # Convert the raw sequence string into a file-like object\n",
        "    sequence_file_like = StringIO(sequence_string)\n",
        "    records = list(SeqIO.parse(sequence_file_like, \"fasta\"))\n",
        "\n",
        "    # Extract sequences from records using list comprehension\n",
        "    sequences = [str(record.seq) for record in records]\n",
        "\n",
        "    # Use list comprehension to generate kmer counts for each sequence\n",
        "    kmers_lists = [count_kmers(seq, k) for seq in sequences]\n",
        "\n",
        "    total_kmers_count = {}\n",
        "    for kmers_counts in kmers_lists:\n",
        "        for kmer, count in kmers_counts.items():\n",
        "            total_kmers_count[kmer] = total_kmers_count.get(kmer, 0) + count\n",
        "\n",
        "    return total_kmers_count\n",
        "\n",
        "\n",
        "def main():\n",
        "    while True:\n",
        "        print(\"\\nMenu:\")\n",
        "        print(\"1. Count k-mers from sequence\")\n",
        "        print(\"2. Exit\")\n",
        "\n",
        "        choice = input(\"Enter your choice: \")\n",
        "\n",
        "        if choice == '1':\n",
        "            try:\n",
        "                sequence_string = download_sequence()\n",
        "                k = int(input(\"Please enter the value for k: \"))\n",
        "\n",
        "                if k < 1:\n",
        "                    raise ValueError(\"k must be a positive integer.\")\n",
        "\n",
        "                result = count_kmers_from_string(sequence_string, k)\n",
        "                print(result)\n",
        "            except ValueError as ve:\n",
        "                print(f\"Error: {ve}\")\n",
        "            except Exception as e:\n",
        "                print(f\"An unexpected error occurred: {str(e)}\")\n",
        "        elif choice == '2':\n",
        "            print(\"Exiting...\")\n",
        "            break\n",
        "        else:\n",
        "            print(\"Invalid choice! Please choose again.\")\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#question 2:\n",
        "from Bio import Entrez, SeqIO\n",
        "import numpy as np\n",
        "\n",
        "#question 2\n",
        "\n",
        "\n",
        "def read_fasta_file(file_path):\n",
        "    sequences = {}\n",
        "    for record in SeqIO.parse(file_path, \"fasta\"):\n",
        "        sequence_id = record.id\n",
        "        sequence = str(record.seq)\n",
        "        sequences[sequence_id] = sequence\n",
        "    return sequences\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "# Replace with your email address (required by NCBI)\n",
        "Entrez.email = \"abhinav21440@iiitd.ac.in\"\n",
        "\n",
        "def download_sequence(accession):\n",
        "    try:\n",
        "        # Fetch the sequence from NCBI\n",
        "        handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"gb\", retmode=\"text\")\n",
        "        record = SeqIO.read(handle, \"genbank\")\n",
        "        handle.close()\n",
        "\n",
        "        # Store the sequence in a string variable\n",
        "        sequence = str(record.seq)\n",
        "        return sequence\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred while downloading the sequence: {str(e)}\")\n",
        "        return None\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def edit_dist(s_accession_num,t_accession_num):\n",
        "   s=download_sequence(s_accession_num)\n",
        "   t=download_sequence(t_accession_num)\n",
        "  # we first initalize a matrix of m+1, n+1 size filled with zeros\n",
        "   row=len(s)\n",
        "   col=len(t)\n",
        "\n",
        "   dp= np.zeros((row+1,col+1), dtype=int)\n",
        "\n",
        "\n",
        "  # now in edit distance we initalize the first row and first col as 0 as gap penalty is zero\n",
        "\n",
        "  # for first col\n",
        "\n",
        "   # finding the edit distance using the matrix\n",
        "\n",
        "   for i in range(row+1):\n",
        "     for j in range(col+1):\n",
        "\n",
        "      if(i==0):\n",
        "        dp[i][j]=j\n",
        "\n",
        "      elif(j==0):\n",
        "         dp[i][j]=i\n",
        "\n",
        "         #this is in the case of a match\n",
        "\n",
        "      elif(s[i-1]==t[j-1]):\n",
        "          dp[i][j]=dp[i-1][j-1]\n",
        "      # these are the\n",
        "      else:\n",
        "        deletions=dp[i-1][j]\n",
        "        insertions= dp[i][j-1]\n",
        "        substitutions= dp[i-1][j-1]\n",
        "\n",
        "        dp[i][j]= 1+ min(deletions, insertions, substitutions)\n",
        "\n",
        "   return dp[row][col]\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "#the initial accession numbers of the sequences\n",
        "s_accession_number = \"NM_001301717\"\n",
        "t_accession_number = \"NM_001001755\"\n",
        "\n",
        "print(\"The edit distance between s and t is \", edit_dist(s_accession_number,t_accession_number))\n"
      ],
      "metadata": {
        "id": "Tbdkg5a_-Qmv",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e7686bca-85f1-46e9-ccb1-0724f3fc4ad9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The edit distance between s and t is  1880\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#question 3:\n",
        "from Bio import Entrez, SeqIO\n",
        "\n",
        "# Replace with your email address (required by NCBI)\n",
        "Entrez.email = \"shivam21489@iiitd.ac.in\"\n",
        "\n",
        "def download_sequence(accession):\n",
        "    try:\n",
        "        # Fetch the sequence from NCBI\n",
        "        handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"gb\", retmode=\"text\")\n",
        "        record = SeqIO.read(handle, \"genbank\")\n",
        "        handle.close()\n",
        "\n",
        "        # Store the sequence in a string variable\n",
        "        sequence = str(record.seq)\n",
        "        return sequence\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred while downloading the sequence: {str(e)}\")\n",
        "        return None\n",
        "\n",
        "def reconstruct_alignment(s, t, n, m, dp):  # function which calculates the reconstruct the alignment\n",
        "    aligned_s, aligned_t = \"\", \"\"  # Initialize aligned sequences as empty strings\n",
        "    i, j = n, m  # Initialize indices for sequence s and t to their lengths\n",
        "\n",
        "    # Start a loop to reconstruct the alignment\n",
        "    while i > 0 or j > 0:\n",
        "        # Check if we can match or mismatch a character\n",
        "        if i > 0 and j > 0 and (s[i - 1] == t[j - 1] or dp[i][j] == dp[i - 1][j - 1] + 1):\n",
        "            # Characters at positions i-1 in s and j-1 in t match or have been substituted\n",
        "\n",
        "            aligned_s = s[i - 1] + aligned_s\n",
        "            aligned_t = t[j - 1] + aligned_t\n",
        "            i -= 1  # Move to the previous positions in both s and t\n",
        "            j -= 1\n",
        "        # Check if we need to insert a gap in s\n",
        "        elif i > 0 and dp[i][j] == dp[i - 1][j] + 1:\n",
        "\n",
        "            aligned_s = s[i - 1] + aligned_s\n",
        "            aligned_t = \"-\" + aligned_t\n",
        "            i -= 1  # Move to the previous position in s\n",
        "        else:\n",
        "           # Check if we need to insert a gap in t\n",
        "\n",
        "            aligned_s = \"-\" + aligned_s\n",
        "            aligned_t = t[j - 1] + aligned_t\n",
        "            j -= 1  # Move to the previous position in t\n",
        "\n",
        "\n",
        "# reconstructs the alignment by considering matches, substitutions, insertions, and deletions\n",
        "\n",
        "    return dp[n][m], aligned_s, aligned_t  # Return the edit distance and the aligned sequences\n",
        "\n",
        "\n",
        "def global_alignment(s_accession, t_accession):\n",
        "    s = download_sequence(s_accession)\n",
        "    t = download_sequence(t_accession)\n",
        "    #print(len(s))\n",
        "    #print(len(t))\n",
        "    print(f\"length of the nucleotide sequence s: {len(s)}\")\n",
        "    print(f\"length of the nucleotide sequence t: {len(t)}\")\n",
        "    if s is None or t is None:\n",
        "        return None, None, None\n",
        "\n",
        "    n, m = len(s), len(t)\n",
        "    # initializes a 2D list dp with dimensions (n + 1) x (m + 1) and fills it with zeros\n",
        "    dp = []\n",
        "    for i in range(n + 1):\n",
        "      row = []\n",
        "      for j in range(m + 1):\n",
        "          row.append(0)\n",
        "      dp.append(row)\n",
        "\n",
        "    for i in range(1, n + 1):\n",
        "        for j in range(1, m + 1):\n",
        "           match = dp[i - 1][j - 1] if s[i - 1] == t[j - 1] else dp[i - 1][j - 1] + 1\n",
        "           delete = dp[i - 1][j] + 1\n",
        "           insert = dp[i][j - 1] + 1\n",
        "           dp[i][j] = min(match, delete, insert)\n",
        "\n",
        "\n",
        "    return reconstruct_alignment(s, t, n, m, dp)\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "# Define the accession numbers for the two sequences\n",
        "    s_accession_number = \"NM_001301717\"  # accession number for s\n",
        "    t_accession_number = \"NM_001001755\"  # accession number for t\n",
        "\n",
        "# Call the global_alignment function to find the alignment between the two sequences\n",
        "    result = global_alignment(s_accession_number, t_accession_number)\n",
        "\n",
        "# Check if the alignment was successful (result is not None)\n",
        "    if result is not None:\n",
        "\n",
        "        distance, s_prime, t_prime = result\n",
        "# Print the edit distance and the aligned sequences\n",
        "        print(f\"Distance: {distance}\")\n",
        "        print(f\"s_prime: {s_prime}\")\n",
        "        print(f\"t_prime: {t_prime}\")\n",
        "\n",
        "# If the alignment failed, print an error message\n",
        "    else:\n",
        "        print(\"Failed to download one or both sequences.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "voJAu_irFMsl",
        "outputId": "43f138e9-fff1-46ea-9a04-322c99531823"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "length of the nucleotide sequence s: 2191\n",
            "length of the nucleotide sequence t: 3537\n",
            "Distance: 1112\n",
            "s_prime: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------CTCTAGATGAGTCAGTGGAGGGCGGGTGGAGCGTTGAACCGTGAAGAGTGTGGTTGGGCG-TA-AACGTGGACTTAAACTCA-GGAGCTAAGGGGGAAACCAATGAAAAGCGTGCTGGTGGTGGCTCTCCTTGTCATTTTCCAGGTATGCCTGTGTCAAGATGAGGTCACGGACGATTACATCGGAGACAACACCACAGTGGACTACACTTTGTTCGAGTC-TTT--GTGCTCC-AAGAAGGACGTGCGGAACTTTAAAGCCTGGTTCCTCCCTATCATGTACTCCATCATTTGTTTCGTGGGCCTACTGGGCAATGGGCTGGTCGTGTTGACCTATATCTATTTCAAGAGGCTCAAGACCATGACCGATACCTACCTGCTCAACCTGGCGGTGGCAGACATCCTCTTCCTCCTGACCCTTCCCTTCTGGGCCTACAGCGC-GGCCAAGTCCTGGGTCTTCGGTGTCCACTTTTGCAAGCTCATCTTTGCCATCTACAAGATGAGCTTCTTCAGTGGCATGCTCCTACTTCTTTGCATCAGCATTGACCGCTACGTGGCCATCGTCCAGGCTGTCTCAGCTCACCGCCACCGTGCCCGCGTCCTTCTCATCAGCAAGCTGTCCTGTGTGGGCATCTG-GATACTAGCCACAGTGCTCTCCATCCCAGAGCTCCTGTACAG--TGACCTCCAGAGGAGCAGCAGTGAGCAAGCGATGCGATGCTCTCTCATCACAGAGCATGTGGAGGCCTTTATCACCATCCAGGTGGCCCAGATGGTGATCGGCTTTCTGGTCCCCCTGCTGGCCATGAGCTTCTGTTACCTTGTCATCATCCGCACCCTGCTCCAGGCACGCAACTTTGAGCGCAACAAGGCCATCAAGGTGATCATCGCTGTGGTCGTGGTCTTCATAGTCTTCCAGCTGCCCTACAATGGGGTGGTCCTGGCCCAGACGGTGGCCAACTTCAAC-ATCACCAGTAGCACCTGTGAGCTCAGTAAGCAACTCAACATCGCCTACGACGTCACCTACAGCCTGGCCTGCGTCCGCTGCTGCGTCAACCCTTTCTTGTACGCCTT-CATCGGCGTCAA-GTTCCGCA-AC-GATC-TCTTCAAGCTCTTCAA--GGA--CCTGGGCTGCCTCAGCCAGGAGCAGCTCCGGCAGTGGTCTTCCTGTCGGCACATCCGGCGCTCCTCCATGAGTGTGGAGGCCGAGACCACCACCACCTTCTCCCCATAGGCGACTCTTCTGCCTG--GACTAGAGG-GACCTCTCCCAGGGTCCCTGGGGTGGGGATAGGGAGCAGATGCAATGACTCAGGACATCCCCCCGCCAAAAGCTGCTCAGGGAAAAGCAGCTCTCCCCTCAGAGTGCAAGCCCCTG-CTCCAGAAGATAGCTTCACCCCAATCCCAGCTACCTCAACCAATGCCAAAAAAAGACAGGGCTGATAAGCTAACACCAGACAGACAACACTGGGAAACAGAGGCTATTGTCCCCTAAACCAAAAACTGAAAGTGAA--AGTCCAGAAACTGTTCCCAC-CTGCTGGAGTGAAGGGGCCAAGGAGGGTGAGTGCAAGGGGCGTGGGAGTGGCCTGAAGAGTCCTCTGAATGAACCTTCTGGCCTCCCACAGACTCAAATGCTCAGACCA-GCT--CTTCCGAAAACCAGGCCTTATCTCCAAGACCAG-AGATAGTGGGGAGACTTCTTGGCTTGG-TG-AGGAAAAGCGGACATCAGCTGGTCAAACAAACTCTCTGAACCCCTCCCTCCATCG-TTTTCTTCACTGTC-CTCCAAGCCAGCGGGAATGGCAGCTGCCACGCCGCCCTAAAAGCACACTCATCCCCTCACTTGCCGCGTCGCCCTCCCAGGCTCTCAACAGGGGAGAGTGTGGTGTTTCCTGCAGGCCAGGCCAGCTGCCTCCGCGTGAT-CAAAGCCACACTCTGGGCTCCAGAGTGGGGATGACATGCACTCAGCT--C-TTGGCTCCACTGGGATGGGAGGAGAGGACAAGGGAAATGTCAGGGGCGGGGAGGGTGACAGTGGCCGCCCAAGGCCCACGAGCTTGTTCTTTGTTCTTTGTCACAGGGACTGAAAACCTCTCCTCATGTTCTGCTTTCGATTCGTTAAGAGAGCAACATTTTACCCACACACAGATAAAGTTTTCCCTTGAGGAAACAACAGCTTTAAAAGAAAAAGAAAAAAAAAGTCTTTGGTAAATGGCAAA\n",
            "t_prime: ATGCTGCAGAGGAGCAGGCTGTTGTGGCTTGCTGTCTTCATAACCCTGTGGGTTTCCTCAGATGCTCAGGATGATGCCAAAGAAGAGGAGAATACCTTTGATTTACTTCAAATCAGTAACATAAACCGAAAGACAATTGGAGCAAAATTGTTTCGGGGTCCAGATCCCGCTATCCCAGCATATCGTTTCATTCGGTTTGACCATATACCTCCATTTAAACCAGAGAAGTTAAAAAAGATTGTTAAACTCATACGGCAAAATGAGGGCTTTATCCTTTCAGCCACCCTGAGACAGGACAGGCAATCTAGAGGCACTATCCTTGCTTTAGAGGGTCCTGGCATCAGTGAGAGGCAGTTTGAGATAATTTCCAATGGCCGAGCCAACACCCTGGACCTCATTTATTGGGTGGATGGCTTCCAGAACGTGATTTCCCTGGAAGACGTGGACCTTGCTGACTCGCAGTGGAAGAACCTCACCGTGCAAGTAACAGGAGAGAACTACAACCTCTATGTTGGCTGTGACCTTATTGACAGCTTCATCCTGGAGGAGCCTTTCTATGAGCAACTGAAAGCAGAGAACAGCAGAATGTATGTGGCGAAAGGGTCCATTCGGGAGAATCATTTCAGGGGTCTTTTGCAAAATATTCATTTAATCTTTGATACTTCAATAGAGGATGTTCTACGTAAAAAAGGATGCCAGAGGAGCCAATCAACTGAAGTGAACACCATCAATGAGAGCACTGAGATTCTTCACTTGAGCCCTGCTGTCACGACTGAGTATGTAGGTGAAAAGACAGAGAAGAAGGCTGAGTTCTGTGACCGCTCTTGTGAAGAGCTGGGAACAATGTTCACTGAACTCACTGGCCTGCGCATTGTGGTGAACAATTTAGCTGATAACCTCCAAAAAGTGTCTGAAGAGAACCAAATAATGTGGGAATTGATTGGCCCTAACAAAACACTCAAGAACCAATCAGTTTGCTGGCAGGATGGCCGAGTCTTTGCAGATAGTGAAAGCTGGATTGTAGATAGCTGCACAAAGTGCACCTGCCAGGACTCTAAAATTGTGTGCCATCAAATCACCTGTCCACCAGTTTCTTGTGCTGATCCGTCCTTTATTGAAGGTGAATGCTGTCCAGTCTGCTCTCACTCTGATGATAGTGAGGAAGGCTGGTCACCCTGGTCAGACTGGACAAAGTGCTCAGTTACCTGTGGCTCTGGGACTCAGATGAGAGGAAGGTCTTGTGATGTCACCAGGAGTGCTTGTACAGGCCCACACATTCAGACGAGGATGTGCAGCTTCAAGAAATGTGACCATCGCATACGTCAAGATGGTGGTTGGAGTCACTGGTCTCCCTGGTCCTCATGTTCAGTTACCTGTGGAGTAGGAAATATCACCCGCATCCGCCTCTGCAATTCCCCTATTCCCCAGATGGGTGGGAAAAACTGTGTGGGAAATGGACGTGAAACAGAGAAATGTGAGAAGGCCCCGTGCCCAGTGAATGGCCAGTGGGGTCCTTGGTCTCCATGGTCTGCTTGCACTGTTACCTGTGGAGGAGGAATCCGTGAGAGGTCTCGCCTCTGTAACAGTCCAGAGCCACAGTATGGAGGAAAGCCATGTGTGGGAGATACCAAGCAACATGATATGTGCAATAAAAGGGATTGCCCAATTGATGGTTGTTTGTCAAATCCTTGCTTTCCTGGAGCAGAATGCAACAGTTACCCGGATGGATCGTGGTCATGTGGGCCATGCCCAGCAGGTTTCCTTGGCAATGGTACAGTCTGTGAGGATCTTGATGAGTGTATAGCTGTGTCAGATGTTTGCTTTAAGGTGAATCAGGTCCATCGCTGT-G-TGA-ACA--CAAATCCGGGT-TTCCATTG---CTTG-CCA-TGT-CCTCCGCGTTATAAAG-GAAGTCAGCCTTACGGAG-TAGGGCTGGAA--GTTGCCAA--G-AC-AG-AGAAGCAAGTC-TGTGAACCT-GA-GAAT-CC-ATG-CAA-A-GA---CA-AGAC-A-CACAGCTGCCATAA-A--TCTGCAGAGTGCA-TTTATCTGGGTCATTTCAGTGATCCTATGTA-CAAATG-TGAA-TGT-AGGACAGGTT-ATGCTGGTGATGGAC-GCATC---TG--TGGTGAG--GATTCAG-ATTTGGATGG--ATG--G-CC-A-A-ATA-AT-AACTTGGT-TTGTGC-TG-CTAAT-GCTACGT-ATC-A-CT-G-TGT-GAAG-GA------T--AACTG--CC--CCCTCCT--GCCTA-A-CTCTGGCCAAG----AAGACTT---TGATAAAGATGGCAA-AGGA--GATGCC-TGT----GATGAG---GATGA-TGACA--ATGAT-GGT-GTTG-A--AG-A-TGA---C-A-AAGACAATTG-CC---CTCT-TCTGTTCA-ATCC-TCGT--CAGTTTGATTATGA-CA--AAGATGAAGT-TGGTGACCGCTGTGATAATTGCC-C--TTATGTGCA---CA-A-C-CCCGCACAGATTGA-C-ACAGA-TA--A-TAATG-G---G-GA-G-GGTG--AT-TCAT----GTGC-TGTGGA------TAT-----T---GATGG---AGA---TGA-C--ATTT-T--T---AATGAACGAGATAATTGTC-CTTACGTT-T-ATAAT----A--CAGAT-CA-G-A-G------TGA----TAC-A-G--AT---GGTG---AT-G--G-AGT--TGGTGATC--AGTGTGATAACTGTCCGCTAAT-GCATAATCC-AGACCAGACTGATG-C-AGAT-AACGAT--CTTGTTG-----GTGA-C-CAGT--GTGA--CAACAATG---AGGACAT-AGATGAAG-ATGGCC-AC--CAG-AACAAC--CAA-GATAAC-TG--C-CCTTACATC--C-CCAATGCTAACCAGGCTGACCATGATAAAGATGGTAAAGGGGATGCCTGTGAT-CCTGA-TGATGA-CA-AT---G-A-TGGTATTCCAG-ATG-ACA---GG-G-----ACA--A-T-TGTCGCCTCAGA-TA-CA--ACC--CTGAGCA-GGAGGACTCTGATG-GTGATGGC-AGAGGTGATATTTGCAAAGATGACT---TTGACGAT---GA-CA-ATG---T-TC-CA-GATAT--TTTTG---ATGTATG-TC-CTGAAAA-CA---ATGCCATC--AGTG-AA---ACTGATTTCAG-A-A-AG-TT---CCAGAT----GGT-CC-C--TC--TG------GACCCCAAGG--G--AA-C-AGC-TCAGATTG--ATC-C----CAACTG-GGTTA-T-TCGCC---ACC---AA-GGCAAG-GAACTGGTGCAG--ACTG--CCAACTCTGATCCTG-GAA-TAG-C-TGTA-GGTTA-TG--A--TGAGT-TCA--GCTCTGTTGA--CTTCAG-TGGCACATTCTATGTTAACACAGACCGTGATGAT--GACTATGCTGGCTT-TG--TTTTTGG-C-TA---CC-AGTCCAGCAG-CCGTTTTTA-TGTTC-TGATGTGGAAGCAGGTCACTCAGACCT-A-CTGGGAGGACAAAC-C-C---A-CCAGGGCT-TATGGCTATTCTGGTGTGTCACT-CAAGGTA--GTGAAT-TCA---ACAAC-TGGTACT-GGGGAACACT--T--GC-GAAATG-CTCTTTGGCAT-ACAGG-AAACACCCCTGGACA----GGTG----C-GCA---C--TTTA--TG-----GCATGATCCTAAG-AACA-T-T-GGCT--GGAAAGATTAT-AC-TGCATACAGGTGGCATTTGATTCACAGGCCT---AAAACAGGACTA-ATAAA----A-GTTC-TAGTGTATGA-AG-GGAAG--CAA-G-TCATG-G-TGGAT-TCTGGAC-CAATC-TA-TGA-T-ACAA-C-CT--T--TG--CTG--GTGGA--CGATTAG-G-TC--TA-TTTGTCTTCTCTCA-AGAAA-TGGT--C-T-A--TTTC-TCTGACCT--CA-AATATG-AATGCAGAG---ATGCT---T-G---A\n"
          ]
        }
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}